hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java

```diff
@@ -3736,4 +3741,67 @@ public CompactedHFilesDischarger getCompactedHFilesDischarger() {
   public long getRetryPauseTime() {
     return this.retryPauseTime;
   }
+
+  @Override
+  public boolean isCompactionOffloadEnabled(){
+    return regionServerCompactionOffloadManager.isCompactionOffloadEnabled();
+  }
+
+  private synchronized void createCompactionManagerStub(boolean refresh) {
+    // Create Master Compaction service stub without refreshing the master node from ZK,
+    // use cached data
+    if (cmsStub == null) {
+      cmsStub =
+          (CompactionService.BlockingInterface) createMasterStub(CompactionService.class, refresh);
+    }
+  }
+
+  /**
+   * Send compaction request to compaction manager
+   * @return True if send request successfully, otherwise false
+   * @throws IOException If an error occurs
+   */
+  @Override
+  public boolean requestCompactRegion(RegionInfo regionInfo, ColumnFamilyDescriptor cfd,
+    boolean major, int priority) {
+    if (!isCompactionOffloadEnabled()) {
+      return false;
+    }
+    if (cmsStub == null) {
+      createCompactionManagerStub(false);
```

Comment:
Yes, we can global control the throttle, the compaction task to compaction server mapping and so on

Reply:
without master, we can't send compact request to compaction server. However, we can do compact in regionserver as original

Reply:
without master, we can't send compact request to compaction server. However, we can do compact in regionserver as original

Reply:
So the logic is already implemented? I saw in the last catch block, we will return false, which is the same return value with compaction offload disabled.

This will be a challenge of the CP design, as for the same table, the compaction could be run in a compaction server or the region server randomly... Anyway, not the problem for this patch, but we need to consider again.